src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/WorkspaceDiagnosticsRefresher.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.LanguageServer.Hosting;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer;
+
+internal sealed class WorkspaceDiagnosticsRefresher : IRazorStartupService
+{
+    private readonly AsyncBatchingWorkQueue _queue;
+    private readonly IClientCapabilitiesService _clientCapabilitiesService;
+    private readonly IClientConnection _clientConnection;
+    private bool? _supported;
+
+    public WorkspaceDiagnosticsRefresher(
+        IProjectSnapshotManager projectSnapshotManager,
+        IClientCapabilitiesService clientCapabilitiesService,
+        IClientConnection clientConnection)
+    {
+        _clientConnection = clientConnection;
+        _clientCapabilitiesService = clientCapabilitiesService;
+        _queue = new(
+            TimeSpan.FromMilliseconds(200),
+            ProcessBatchAsync,
+            default);
```

Comment:
Please consider making this service disposable and add a CancellationTokenSource that is cancelled on Dispose() and its token is passed in here.

As an aside, it's best to be specific and always pass `CancellationToken.None` for cancellation tokens rather than `default`. That way it's easier to use FAR to find places where a cancellation token was forgotten.